internal/testrunner/runners/system/servicedeployer/factory.go

```diff
@@ -5,27 +5,66 @@
 package servicedeployer
 
 import (
-	"errors"
 	"os"
 	"path/filepath"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/elastic-package/internal/logger"
 )
 
+const devDeployDir = "_dev/deploy"
+
 var (
 	// ErrNotFound is returned when the appropriate service runner for a package
 	// cannot be found.
 	ErrNotFound = errors.New("unable to find service runner")
 )
 
-// Factory chooses the appropriate service runner for the given package, depending
-// on service configuration files defined in the package.
-func Factory(packageRootPath string) (ServiceDeployer, error) {
-	packageDevPath := filepath.Join(packageRootPath, "_dev")
+// FactoryOptions defines options used to create an instance of a service deployer.
+type FactoryOptions struct {
+	PackageRootPath    string
+	DataStreamRootPath string
+}
+
+// Factory chooses the appropriate service runner for the given data stream, depending
+// on service configuration files defined in the package or data stream.
+func Factory(options FactoryOptions) (ServiceDeployer, error) {
+	devDeployPath, err := findDevDeployPath(options)
+	if err != nil {
+		logger.Errorf("can't find _dev/deploy directory")
+		return nil, ErrNotFound
+	}
 
 	// Is the service defined using a docker compose configuration file?
-	dockerComposeYMLPath := filepath.Join(packageDevPath, "deploy", "docker", "docker-compose.yml")
+	dockerComposeYMLPath := filepath.Join(devDeployPath, "docker", "docker-compose.yml")
 	if _, err := os.Stat(dockerComposeYMLPath); err == nil {
 		return NewDockerComposeServiceDeployer(dockerComposeYMLPath)
 	}
 
+	// Is the service defined using Terraform definition files?
+	terraformDirPath := filepath.Join(devDeployPath, "tf")
```

Comment:
With the addition of the terraform deployer and the logic in this `Factory` function, we now have a precedence ordering. If a data set somehow defines folders for both types of deployers we will choose the docker compose one. Maybe this is something we should document in the system testing HOWTO document?

Reply:
I described the change in docs (in this PR) and also described a bit the terraform service deployer (but without diving into details).

Reply:
Hmm, maybe I missed it but I'm not seeing anything in the docs in this PR that talk about the precedence ordering between the service deployers. I'm looking for something in the docs that would explain that only one service deployer should be defined for a package and if multiple ones are defined the Docker Compose service deployer will take precedence (or we can say the results are unpredictable).

Reply:
How about also adding a check in this code to see if more than one service deployer is defined and then return an error in that case?

Reply:
I admit I forgot about his idea. Let me quickly implement this one.

Reply:
Hahaha, seems like I forgot about it too 😄. Thanks for finding it again — just re-request my review on the PR as usual when you have this implemented. Thanks!

Reply:
Ok, now it's fixed :D